server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportUpdateDesiredNodesAction.java

```diff
@@ -174,11 +187,74 @@ public ClusterState execute(ClusterState currentState, List<TaskContext<UpdateDe
                 }
                 final var replacedExistingHistoryId = previousDesiredNodes != null
                     && previousDesiredNodes.hasSameHistoryId(desiredNodes) == false;
-                taskContext.success(
-                    () -> taskContext.getTask().listener().onResponse(new UpdateDesiredNodesResponse(replacedExistingHistoryId))
-                );
+                pendingTasks.add(new PendingTask(taskContext, replacedExistingHistoryId));
+            }
+
+            final var updatedClusterState = DesiredNodes.updateDesiredNodesStatusIfNeeded(currentState, desiredNodes);
+            if (updatedClusterState == currentState) {
+                for (final var pendingTaskListener : pendingTasks) {
+                    pendingTaskListener.runOnPublicationSuccess(pendingTaskListener::notifySuccessfulUpdateToListener);
+                }
+                return currentState;
+            } else {
+                // Trigger a reroute once the desired nodes updates have been published
+                final GroupedActionListener<Void> publicationListener = new GroupedActionListener<>(new ActionListener<>() {
+                    @Override
+                    public void onResponse(Collection<Void> unused) {
+                        rerouteService.reroute("upgraded desired nodes", Priority.URGENT, new ActionListener<>() {
+                            @Override
+                            public void onResponse(ClusterState clusterState) {
+                                pendingTasks.forEach(PendingTask::notifySuccessfulUpdateToListener);
+                            }
+
+                            @Override
+                            public void onFailure(Exception e) {
+                                if (MasterService.isPublishFailureException(e)) {
```

Comment:
> It could simplify all of this a lot if we just succeeded all task-contexts and overrode clusterStatePublished on the executor to do the reroute, what would go wrong if we did this?

I agree, this is a bit complex but it has the nice property of notifying the update desired nodes listeners _after_ the reroute finishes. It's true that this is not super important for orchestrators though, it mostly simplifies our testing code. I'll take your approach.